src/CoreMedia/CMFormatDescription.cs

```diff
@@ -374,7 +461,132 @@ internal CMAudioFormatDescription (IntPtr handle, bool owns)
+		{
+		}
+
+		public CMAudioFormatDescription (ref AudioStreamBasicDescription asbd, AudioChannelLayout layout, NSDictionary extensions) : this (ref asbd, layout, 0, IntPtr.Zero, extensions)
+		{
+		}
+
+		public CMAudioFormatDescription (ref AudioStreamBasicDescription asbd, uint magicCookieSize, IntPtr magicCookie, NSDictionary extensions) : this (ref asbd, null, magicCookieSize, magicCookie, extensions)
+		{
+		}
+
+		public CMAudioFormatDescription (ref AudioStreamBasicDescription asbd, AudioChannelLayout layout, uint magicCookieSize, IntPtr magicCookie, NSDictionary extensions)
+		{
+			IntPtr ptr = IntPtr.Zero;
+			uint size = 0;
+			if (layout != null) {
+				size = (uint)Marshal.SizeOf (layout.NativeStruct);
+				Marshal.StructureToPtr (layout.NativeStruct, ptr, false);
+			}
+			var error = CMAudioFormatDescriptionCreate (IntPtr.Zero, ref asbd, size, ptr, magicCookieSize, magicCookie, extensions != null ? extensions.Handle : IntPtr.Zero, out handle);
+			
+			if (error != CMFormatDescriptionError.None)
+				throw new ArgumentException (error.ToString ());
+		}
+
+		public unsafe AudioStreamBasicDescription? StreamBasicDescription {
+			get {
+				var ret = CMAudioFormatDescriptionGetStreamBasicDescription (handle);
+				if (ret != IntPtr.Zero) {
+					return *((AudioStreamBasicDescription*)ret);
+				}
+				return null;
+			}
+		}
+
+		public AudioChannelLayout ChannelLayout {
+			get {
+				nint size;
+				var res = CMAudioFormatDescriptionGetChannelLayout (handle, out size);
+				if (res == IntPtr.Zero || size == 0)
+					return null;
+				return AudioChannelLayout.FromHandle (res);
+			}
+		}
+
+		public AudioFormat [] Formats {
+			get {
+				unsafe {
+					nint size;
+					var v = CMAudioFormatDescriptionGetFormatList (handle, out size);
+					if (v == IntPtr.Zero)
+						return null;
+					var items = size / sizeof (AudioFormat);
+					var ret = new AudioFormat [items];
+					var ptr = (AudioFormat *) v;
+					for (int i = 0; i < items; i++)
+						ret [i] = ptr [i];
+					return ret;
+				}
+			}
+		}
+
+		public byte [] MagicCookie {
+			get {
+				nint size;
+				var h = CMAudioFormatDescriptionGetMagicCookie (handle, out size);
+				if (h == IntPtr.Zero)
+					return null;
+
+				var result = new byte [size];
+				Marshal.Copy (h, result, 0, result.Length);
+				return result;
+			}
+		}
+
+		public AudioFormat MostCompatibleFormat {
+			get {
+				unsafe {
+					var ret = (AudioFormat *) CMAudioFormatDescriptionGetMostCompatibleFormat (handle);
+					if (ret == null)
+						return new AudioFormat ();
+					return *ret;
+				}
+			}
+		}
+
+		public AudioFormat RichestDecodableFormat {
+			get {
+				unsafe {
+					var ret = (AudioFormat *) CMAudioFormatDescriptionGetRichestDecodableFormat (handle);
+					if (ret == null)
+						return new AudioFormat ();
```

Comment:
From the old PR:

rolfbjarne added a note on Feb 15
Maybe it would be better to return a nullable type instead (AudioFormat?)?
 
timrisi added a note on Feb 16

For this one, it could be changed to return a nullable type if desired. I copied it from the AudioMostCompatibleFormat property in the base class (that should have been put in CMAudioFormatDescription in the first place). I can't change the original, but I could make the new one nullable (at the cost of working marginally differently if anyone was actually using the old one)

timrisi added a note on Feb 16
The old version will still be around until th enext breaking change for XAMCORE_4_0, so probably not the end of the world ifw e do want to change the behavior with this one.
 
rolfbjarne added a note on Feb 16
@spouliot what do you think about this one? IMHO a nullable property looks a lot nicer that returning an empty struct.